### PR TITLE
Feature/2/3 use real tx_hash and merge with session

### PR DESCRIPTION
**Pull request by piniom**
In this PR I've added the logic required for the transaction hash to be verified correctly. This took a bit longer than expected due to some ghosts in the alexandria code responsible for the base64 encoding. For most (although not all) felt252 values sufficiently large the code panicked unexpectedly. I've run generative tests but was unable to reproduce the error when used outside of a contract. I've decided to rewrite the encoding myself. The new encoding is more efficient and as a bonus doesn't panic unexpectedly :)). 
I've created a PR in alexandria which I hope will be merged soon:
https://github.com/keep-starknet-strange/alexandria/pull/237
In the meantime, required parts of the code are included here. 
I've also added logic to support multiple authorization methods so that both webauthn and session can be used as well as the starknet pair. 
Next I'll add support to store multiple keys, create snake_case interfaces and rename functions to be compatible with starknet conventions.

### CODE REVIEW

File: cartridge_account/src/lib.cairo [hunk 1]
```diff
@@ -213,11 +204,28 @@ mod Account {
...
+            let tx_hash = tx_info.transaction_hash;
+            let mut signature = tx_info.signature;
+            if signature.len() == 2_u32 {
+                return self.validate_ecdsa_transaction();
+            } 
+            let signature_type = *signature.at(0_u32);
+            if signature_type == 'Session Token v1' {
+                SessionImpl::validate_session_serialized(self, self.get_public_key(), signature, calls.span())
+            } else if signature_type == 'Webauthn v1' {
+                WebauthnImpl::verifyWebauthnSignerSerialized(self, signature, tx_hash)
```
Comment (glihm) [inline cartridge_account/src/lib.cairo]: Would be great to have the signature types in an enum or constants.

File: src/auth/src/component.cairo [hunk 2]
```diff
@@ -31,14 +32,20 @@ trait IWebauthn<TContractState> {
     fn verifyWebauthnSigner(
         self: @TContractState, 
         signature: WebauthnSignature,
-        tx_hash: Array<u8>,
+        tx_hash: felt252,
     ) -> bool;
+    fn verifyWebauthnSignerSerialized(
```
Comment (glihm) [inline src/auth/src/component.cairo:37]:
```suggestion
    fn verify_webauthn_signer_serialized(
```

File: cartridge_account/src/lib.cairo [hunk 3]
```diff
@@ -229,29 +237,15 @@ mod Account {
             self.emit(OwnerAdded { new_owner_guid: new_public_key });
         }
 
-        fn _is_valid_signature(
+        fn _is_valid_ecdsa_signature(
```
Comment (glihm) [inline cartridge_account/src/lib.cairo]: Leading underscore may be removed as commented in earlier PR.

File: account_sdk/src/webauthn_signer/mod.rs [hunk 4]
```diff
@@ -16,6 +16,8 @@ pub mod credential;
 pub type U256 = (FieldElement, FieldElement);
 pub type Secp256r1Point = (U256, U256);
 
+pub const WEBAUTHN_SIGNATURE_TYPE: FieldElement = felt!("0x576562617574686e207631");
```
Comment (glihm) [inline account_sdk/src/webauthn_signer/mod.rs]:
```suggestion
// "Webauthn v1"
pub const WEBAUTHN_SIGNATURE_TYPE: FieldElement = felt!("0x576562617574686e207631"); 
```

File: src/auth/src/component.cairo [hunk 5]
```diff
@@ -103,6 +114,19 @@ mod webauthn_component {
                 signature.authenticator_data
             ).is_ok()
         }
+        fn verifyWebauthnSignerSerialized(
```
Comment (glihm) [inline src/auth/src/component.cairo:117]:
```suggestion
        fn verify_webauthn_signer_serialized(
```

**Review by glihm — COMMENTED** [review]
> I've decided to rewrite the encoding myself. The new encoding is more efficient and as a bonus doesn't panic unexpectedly :)).

:crown:

File: cartridge_account/src/lib.cairo (hunk 1 above)
Comment (piniom) [reply cartridge_account/src/lib.cairo]: Done :)

File: cartridge_account/src/lib.cairo (hunk 3 above)
Comment (piniom) [reply cartridge_account/src/lib.cairo]: Sorry, I got confused with the branches :). This is fixed in the subsequent PR: [feature/2/5](https://github.com/neotheprogramist/cairo-webauthn/pull/28)

File: account_sdk/src/webauthn_signer/mod.rs (hunk 4 above)
Comment (piniom) [reply account_sdk/src/webauthn_signer/mod.rs]: Same: [feature/2/5](https://github.com/neotheprogramist/cairo-webauthn/pull/28)

**Pull request merged**